Misleading `No exception was actually thrown, this exception is used purely to log trace` and useless trace to `LoggerModel.setup` in logs

**Description**

Misleading `No exception was actually thrown, this exception is used purely to log trace` and useless trace to `LoggerModel.setup` in logs

**To Reproduce**

Steps to reproduce the behavior:

1. Run the 'spring-starter-28' project in IntelliJ Idea
2. Make `application.xml` config malformed by replacing `<bean>` tag with `<beanb>` tag
3. Generate tests using `application.xml` 
4. There's a misleading `No exception was actually thrown, this exception is used purely to log trace` message in `rdEngineProcessLogs` right after actual exception trace:

**Expected behavior**

No `No exception was actually thrown, this exception is used purely to log trace` message

**Actual behavior**

`No exception was actually thrown, this exception is used purely to log trace` message right after actual exception trace

**Visual proofs (screenshots, logs, images)**

```
14:28:48.783 | ERROR | SpringAnalyzerProcess     | RdCategory: SpringApplicationAnalyzerFacade | Analyzer org.utbot.spring.analyzers.PureSpringApplicationAnalyzer@7db85202 failed | org.springframework.beans.factory.xml.XmlBeanDefinitionStoreException: Line 50 in XML document from class path resource [application.xml] is invalid; nested exception is org.xml.sax.SAXParseException; lineNumber: 50; columnNumber: 49; cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://www.springframework.org/schema/beans":beanb}'. One of '{"http://www.springframework.org/schema/beans":import, "http://www.springframework.org/schema/beans":alias, "http://www.springframework.org/schema/beans":bean, WC[##other:"http://www.springframework.org/schema/beans"], "http://www.springframework.org/schema/beans":beans}' is expected.
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:402)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:338)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:310)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:188)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:224)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:195)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.lambda$loadBeanDefinitionsFromImportedResources$0(ConfigurationClassBeanDefinitionReader.java:390)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsFromImportedResources(ConfigurationClassBeanDefinitionReader.java:354)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:156)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:129)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:343)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:247)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:311)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:112)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:746)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564)
	at org.utbot.spring.analyzers.PureSpringApplicationAnalyzer$analyze$1.invoke(PureSpringApplicationAnalyzer.kt:12)
	at org.utbot.spring.analyzers.PureSpringApplicationAnalyzer$analyze$1.invoke(PureSpringApplicationAnalyzer.kt:12)
	at org.utbot.spring.exception.UtBotSpringShutdownException$Companion.catch(UtBotSpringShutdownException.kt:19)
	at org.utbot.spring.analyzers.PureSpringApplicationAnalyzer.analyze(PureSpringApplicationAnalyzer.kt:12)
	at org.utbot.spring.analyzers.SpringApplicationAnalyzerFacade.analyze(SpringApplicationAnalyzerFacade.kt:28)
	at org.utbot.spring.process.SpringAnalyzerProcessMainKt$setup$1.invoke(SpringAnalyzerProcessMain.kt:50)
	at org.utbot.spring.process.SpringAnalyzerProcessMainKt$setup$1.invoke(SpringAnalyzerProcessMain.kt:43)
	at org.utbot.rd.IdleWatchdog$measureTimeForActiveCall$1$2$1.invoke(ClientProcessUtil.kt:115)
	at org.utbot.rd.IdleWatchdog.wrapActive(ClientProcessUtil.kt:88)
	at org.utbot.rd.IdleWatchdog$measureTimeForActiveCall$1.invoke(ClientProcessUtil.kt:114)
	at com.jetbrains.rd.framework.IRdEndpoint$set$1.invoke(TaskInterfaces.kt:182)
	at com.jetbrains.rd.framework.IRdEndpoint$set$1.invoke(TaskInterfaces.kt:182)
	at com.jetbrains.rd.framework.impl.RdCall.onWireReceived(RdTask.kt:362)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:57)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.impl.ProtocolContexts.readMessageContextAndInvoke(ProtocolContexts.kt:148)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:54)
	at com.jetbrains.rd.util.threading.SingleThreadSchedulerBase.queue$lambda-3(SingleThreadScheduler.kt:41)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.xml.sax.SAXParseException; lineNumber: 50; columnNumber: 49; cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://www.springframework.org/schema/beans":beanb}'. One of '{"http://www.springframework.org/schema/beans":import, "http://www.springframework.org/schema/beans":alias, "http://www.springframework.org/schema/beans":bean, WC[##other:"http://www.springframework.org/schema/beans"], "http://www.springframework.org/schema/beans":beans}' is expected.
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.error(ErrorHandlerWrapper.java:135)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:396)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:284)
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator$XSIErrorReporter.reportError(XMLSchemaValidator.java:512)
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.reportSchemaError(XMLSchemaValidator.java:3600)
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.handleStartElement(XMLSchemaValidator.java:1972)
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.startElement(XMLSchemaValidator.java:830)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:374)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:247)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:342)
	at org.springframework.beans.factory.xml.DefaultDocumentLoader.loadDocument(DefaultDocumentLoader.java:77)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadDocument(XmlBeanDefinitionReader.java:432)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:390)
	... 38 more
 | java.lang.Exception: No exception was actually thrown, this exception is used purely to log trace
	at org.utbot.rd.loggers.UtRdKLogger.format(UtRdKLogger.kt:33)
	at org.utbot.rd.loggers.UtRdKLogger.log(UtRdKLogger.kt:44)
	at org.utbot.rd.loggers.UtRdLogUtilKt$setup$1.invoke(UtRdLogUtil.kt:41)
	at org.utbot.rd.loggers.UtRdLogUtilKt$setup$1.invoke(UtRdLogUtil.kt:38)
	at com.jetbrains.rd.util.reactive.Signal.fire(Signal.kt:32)
	at com.jetbrains.rd.framework.impl.RdSignal.onWireReceived(RdSignal.kt:42)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:57)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.impl.ProtocolContexts.readMessageContextAndInvoke(ProtocolContexts.kt:148)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:54)
	at com.jetbrains.rd.util.threading.SingleThreadSchedulerBase.queue$lambda-3(SingleThreadScheduler.kt:41)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/rd/InstrumentedProcess.kt
+++ b/utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/rd/InstrumentedProcess.kt
@@ -33,7 +33,6 @@ import org.utbot.rd.terminateOnException
 import java.io.File
 
 private val logger = KotlinLogging.logger { }
-private val rdLogger = UtRdKLogger(logger, "")
 
 private const val UTBOT_INSTRUMENTATION = "utbot-instrumentation"
 private const val INSTRUMENTATION_LIB = "lib"
@@ -136,7 +135,7 @@ class InstrumentedProcess private constructor(
             logger.trace("rd process started")
 
             val proc = InstrumentedProcess(classLoader, rdProcess)
-            proc.loggerModel.setup(rdLogger, proc.lifetime)
+            proc.loggerModel.setup(logger, proc.lifetime)
 
             proc.lifetime.onTermination {
                 logger.trace { "process is terminating" }

--- a/utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdKLogger.kt
+++ b/utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdKLogger.kt
@@ -6,7 +6,11 @@ import mu.KLogger
 /**
  * Adapter from RD Logger to KLogger
  */
-class UtRdKLogger(private val realLogger: KLogger, val category: String) : Logger {
+class UtRdKLogger(
+    private val realLogger: KLogger,
+    val category: String,
+    private val logTraceOnError: Boolean = true
+) : Logger {
     val logLevel: LogLevel
         get() {
             return when {
@@ -24,7 +28,11 @@ class UtRdKLogger(private val realLogger: KLogger, val category: String) : Logge
     }
 
     private fun format(level: LogLevel, message: Any?, throwable: Throwable?): String {
-        val throwableToPrint = if (level < LogLevel.Error) throwable else throwable ?: Exception("No exception was actually thrown, this exception is used purely to log trace")
+        val throwableToPrint =
+            if (logTraceOnError && level >= LogLevel.Error)
+                throwable ?: Exception("No exception was actually thrown, this exception is used purely to log trace")
+            else
+                throwable
         val rdCategory = if (category.isNotEmpty()) "RdCategory: ${category.substringAfterLast('.').padEnd(25)} | " else ""
         return "$rdCategory${message?.toString() ?: ""} ${throwableToPrint?.getThrowableText()?.let { "| $it" } ?: ""}"
     }

--- a/utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdLogUtil.kt
+++ b/utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdLogUtil.kt
@@ -26,7 +26,10 @@ fun overrideDefaultRdLoggerFactoryWithKLogger(logger: KLogger) {
     }
 }
 
-fun LoggerModel.setup(rdLogger: UtRdKLogger, processLifetime: Lifetime) {
+fun LoggerModel.setup(logger: KLogger, processLifetime: Lifetime) {
+    // we do not log trace on error here because it's a job of clint-side process logger
+    // server-side process logger can only log trace of where it was set up from which isn't particularly useful
+    val rdLogger = UtRdKLogger(logger, "", logTraceOnError = false)
     // currently we do not specify log level for different categories
     // though it is possible with some additional map on categories -> consider performance
     // this logLevel is obtained from KotlinLogger

--- a/utbot-spring-analyzer/src/main/kotlin/org/utbot/spring/process/SpringAnalyzerProcess.kt
+++ b/utbot-spring-analyzer/src/main/kotlin/org/utbot/spring/process/SpringAnalyzerProcess.kt
@@ -37,7 +37,6 @@ private const val SPRING_ANALYZER_JAR_PATH = "lib/$SPRING_ANALYZER_JAR_FILENAME"
 private const val UNKNOWN_MODIFICATION_TIME = 0L
 
 private val logger = KotlinLogging.logger {}
-private val rdLogger = UtRdKLogger(logger, "")
 
 private var classpathArgs = listOf<String>()
 
@@ -106,7 +105,7 @@ class SpringAnalyzerProcess private constructor(
                 }
                 rdProcess.awaitProcessReady()
                 val proc = SpringAnalyzerProcess(rdProcess)
-                proc.loggerModel.setup(rdLogger, proc.lifetime)
+                proc.loggerModel.setup(logger, proc.lifetime)
                 return proc
             }
     }

